Fix H1 1102067; Fix H1 1102080

aembler opened the pull request.

wordish reviewed (changes requested):
Other than not being very DRY, the only comment I have is that the regex can be simplified to `/[^a-z0-9_-]/i` since the `i` modifier makes it case insensitive.

mlocati commented:
> Other than not being very DRY, the only comment I have is that the regex can be simplified to `/[^a-z0-9_-]/i` since the `i` modifier makes it case insensitive.

We can also use `\w`, whose meaning is "any us-ascii letter (A-Z and a-z), digit (0-9) or underscore (_)": we'd only miss the -

wordish commented:
> > Other than not being very DRY, the only comment I have is that the regex can be simplified to `/[^a-z0-9_-]/i` since the `i` modifier makes it case insensitive.
> 
> We can also use `\w`, whose meaning is "any us-ascii letter (A-Z and a-z), digit (0-9) or underscore (_)": we'd only miss the -

Got to be careful though:

> A "word" character is an underscore or any character that is a letter or digit. By default, the definition of letters and digits is controlled by PCRE's low-valued character tables, and may vary if locale-specific matching is taking place (see "Locale support" in the pcreapi page). For example, in a French locale such as "fr_FR" in Unix-like systems, or "french" in Windows, some character codes greater than 127 are used for accented letters, and these are then matched by \w. The use of locales with Unicode is discouraged. (https://www.pcre.org/original/doc/html/pcrepattern.html)

wordish commented:
I do agree with @mlocati 's comment on another PR that the positive version of this is easier to grok: `/^[a-z0-9_-]+$/i`

mlocati commented:
> For example, in a French locale such as "fr_FR" in Unix-like systems, or "french" in Windows, some character codes greater than 127 are used for accented letters, and these are then matched by \w. 

We don't use single-byte codepages, but UTF-8.
For example, the above sentence could be valid if we used the ISO-8859-1 (aka Latin1) character set, where for example the `à` letter is encoded with  224.
BTW ~concrete5~ Concrete uses UTF-8, which uses multiple bytes to represent the non US-ASCII letters. So, the note above shouldn't be applicable in our case...

wordish commented:
Sure, but there are so many ways this could go wrong if we are relying on this for security. The way regexes were interpreted changed between 7.2 and 7.3, so there's no guarantee that something may happen again in the future, or some quirk of how the server locale, webserver software locale, PHP's encoding settings, the database, and the request all work together to rely on a character class whose meaning does change. If PHP somehow set the `u` modifier to be default, the meaning of `\w` changes. 

Of course, there's no reason why users should not be able to name their files in whatever language they want, but if the current use case is for only English letters, numbers, underscores, and hyphens, it seems better to explicitly state that.

It just seems that validating input should be as explicit as possible.

aembler commented:
I've incorporated the feedback here and updated the regular expression.  I agree with Myq about being as explicit as possible, not just because the \w behavior might change: if it saves a developer even five minutes trying to understand what the code is doing, it seems like it's worth it.

wordish commented:
This will prevent path traversal, so accomplishes the goal. Might be worth revisiting a more systematic approach to validating various types of input sometime, but this looks accepttable to me.

wordish reviewed (approved):
This will prevent path traversal, so accomplishes the goal. Might be worth revisiting a more systematic approach to validating various types of input sometime, but this looks accepttable to me.
